### PR TITLE
add a known limitations section to fake Client

### PR DESCRIPTION
**Pull request by jpatel-pivotal**
This PR adds more information to let users know of the currently known limitations as they exist in `fake.Client`.

We need this to:

* share this list with potential users of the fake client to help with expectations around TDD
* also will help with discussion around which item on the list may be fixed via its own issue. 

This PR also removes the deprecation warning as per issue #768 and is the first step to see if we can improve TDD experience for unit testing by using the fake Client in various testing use cases and scenarios.

### CODE REVIEW

**Comment by jpatel-pivotal** [comment]
I have signed the cla /check-cla

File: pkg/client/fake/doc.go [hunk 1]
```diff
@@ -17,17 +17,24 @@ limitations under the License.
...
 	client := NewFakeClient(initObjs...) // initObjs is a slice of runtime.Object
 
 You can invoke the methods defined in the Client interface.
 
-When it doubt, it's almost always better not to use this package and instead use
+When in doubt, it's almost always better not to use this package and instead use
 envtest.Environment with a real client and API server.
+
+Current Limitations / Known Issues with the fake Client:
+- this client does not use reactors so it can not be setup with detailed responses for testing
```
Comment (vincepri) [inline pkg/client/fake/doc.go:31]:
```suggestion
```

We can probably remove this line, given that it's clear it's a different implementation.

File: pkg/client/fake/doc.go [hunk 2]
```diff
@@ -17,17 +17,24 @@ limitations under the License.
 /*
 Package fake provides a fake client for testing.
 
-Deprecated: please use pkg/envtest for testing. This package will be dropped
-before the v1.0.0 release.
-
-An fake client is backed by its simple object store indexed by GroupVersionResource.
+fake client is backed by its simple object store indexed by GroupVersionResource.
```
Comment (vincepri) [inline pkg/client/fake/doc.go:20]:
```suggestion
A fake client is backed by its simple object store indexed by GroupVersionResource.
```

File: pkg/client/fake/doc.go [hunk 3]
```diff
@@ -17,17 +17,24 @@ limitations under the License.
 /*
 Package fake provides a fake client for testing.
 
-Deprecated: please use pkg/envtest for testing. This package will be dropped
-before the v1.0.0 release.
```
Comment (vincepri) [inline pkg/client/fake/doc.go:21]: cc @DirectXMan12

File: pkg/client/fake/doc.go [hunk 4]
```diff
@@ -17,17 +17,24 @@ limitations under the License.
...
 
 You can invoke the methods defined in the Client interface.
 
-When it doubt, it's almost always better not to use this package and instead use
+When in doubt, it's almost always better not to use this package and instead use
 envtest.Environment with a real client and API server.
+
+Current Limitations / Known Issues with the fake Client:
+- this client does not use reactors so it can not be setup with detailed responses for testing
+- possible locking issues when using fake client and not the actual client
```
Comment (vincepri) [inline pkg/client/fake/doc.go:32]: Which locking issues are we referring to? It'd be great to have more details

File: pkg/client/fake/doc.go [hunk 5]
```diff
@@ -17,17 +17,24 @@ limitations under the License.
...
+- by default you are using the wrong scheme
```
Comment (vincepri) [inline pkg/client/fake/doc.go:33]: What do you mean by this? The function that creates the fake client without a scheme has been deprecated, and will be removed in the next version.

File: pkg/client/fake/doc.go [hunk 6]
```diff
@@ -17,17 +17,24 @@ limitations under the License.
...
+- there is no support for sub resources which can cause issues with tests if you're trying to update
+  e.g. metadata and status in the same reconcile
```
Comment (vincepri) [inline pkg/client/fake/doc.go:35]: There is some support for the `status` subresource

File: pkg/client/fake/doc.go [hunk 7]
```diff
@@ -17,17 +17,24 @@ limitations under the License.
...
+- There is also no OpenAPI validation
```
Comment (vincepri) [inline pkg/client/fake/doc.go:36]:
```suggestion
- No OpeanAPI validation is performed when creating or updating objects.
```

File: pkg/client/fake/doc.go [hunk 8]
```diff
@@ -17,17 +17,24 @@ limitations under the License.
...
+- It does not bump `Generation` or `ResourceVersion` so Patch/Update will not behave properly
```
Comment (vincepri) [inline pkg/client/fake/doc.go:37]:
```suggestion
- ObjectMeta's `Generation` and `ResourceVersion` don't behave properly, Patch or Update operations that rely on these fields will fail, or give false positives.
```

File: pkg/client/fake/doc.go (hunk 4 above)
Comment (jpatel-pivotal) [reply pkg/client/fake/doc.go:32]: I noted this down as we were discussing in the sig meeting so you are right we need more detail on this item. Are there others that can provide more detail here?

File: pkg/client/fake/doc.go (hunk 1 above)
Comment (jpatel-pivotal) [reply pkg/client/fake/doc.go:31]: The use case here is if you would like to test error cases in your controller's reconcile logic, we need a way to inject a specific error. For e.g. to test the psuedo code below:

File: pkg/client/fake/doc.go (hunk 2 above)
Comment (jpatel-pivotal) [reply pkg/client/fake/doc.go]: Fixed

File: pkg/client/fake/doc.go (hunk 5 above)
Comment (jpatel-pivotal) [reply pkg/client/fake/doc.go]: Changed the example in the doc on how to create a new fake Client with a scheme

File: pkg/client/fake/doc.go (hunk 6 above)
Comment (jpatel-pivotal) [reply pkg/client/fake/doc.go]: fixed wording

**Comment by jpatel-pivotal** [comment]
I signed it

File: pkg/client/fake/doc.go (hunk 1 above)
Comment (vincepri) [reply pkg/client/fake/doc.go:31]: Sounds good to me

File: pkg/client/fake/doc.go (hunk 4 above)
Comment (vincepri) [reply pkg/client/fake/doc.go:32]: I'd double check with the recording, but IIRC this is related to locking with update/patch when using resourceVersion, which is covered below

Comment (jpatel-pivotal) [reply pkg/client/fake/doc.go]: You are right. Changed it and resolving this.

File: pkg/client/fake/doc.go [hunk 9]
```diff
@@ -17,17 +17,23 @@ limitations under the License.
...
 You can invoke the methods defined in the Client interface.
 
-When it doubt, it's almost always better not to use this package and instead use
+When in doubt, it's almost always better not to use this package and instead use
 envtest.Environment with a real client and API server.
+
+Current Limitations / Known Issues with the fake Client:
+- this client does not have a way to inject specific errors to test handled vs. unhandled errors
+- there is some support for sub resources which can cause issues with tests if you're trying to update
+  e.g. metadata and status in the same reconcile
```
Comment (vincepri) [inline pkg/client/fake/doc.go:33]:
```suggestion
- This client does not have a way to inject specific errors to test handled vs. unhandled errors.
- There is some support for sub resources which can cause issues with tests if you're trying to update
  e.g. metadata and status in the same reconcile.
```

File: pkg/client/fake/doc.go [hunk 10]
```diff
@@ -17,17 +17,23 @@ limitations under the License.
...
-	client := NewFakeClient(initObjs...) // initObjs is a slice of runtime.Object
+	client := NewFakeClientWithScheme(scheme, initObjs...) // initObjs is a slice of runtime.Object
 
 You can invoke the methods defined in the Client interface.
 
-When it doubt, it's almost always better not to use this package and instead use
+When in doubt, it's almost always better not to use this package and instead use
 envtest.Environment with a real client and API server.
+
+Current Limitations / Known Issues with the fake Client:
```
Comment (vincepri) [inline pkg/client/fake/doc.go:30]:
```suggestion
WARNING: ⚠️ Current Limitations / Known Issues with the fake Client ⚠️
```

**Comment by vincepri** [comment]
@jpatel-pivotal A few other suggestions, otherwise the change should be good to go, @alvaroaleman wdyt

**Pull request merged**